airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,155 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Serialzed DAG table in database."""
+
+import hashlib
+from typing import Any, Dict, List, Optional, TYPE_CHECKING
+from sqlalchemy import Column, Index, Integer, String, Text, and_
+from sqlalchemy.sql import exists
+
+from airflow.models.base import Base, ID_LEN
+from airflow.utils import db, timezone
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+if TYPE_CHECKING:
+    from airflow.dag.serialization.serialized_dag import SerializedDAG  # noqa: F401, E501; # pylint: disable=cyclic-import
+    from airflow.models import DAG  # noqa: F401; # pylint: disable=cyclic-import
+
+
+class SerializedDagModel(Base):
+    """A table for serialized DAGs.
+
+    serialized_dag table is a snapshot of DAG files synchronized by scheduler.
+    This feature is controlled by:
+        [core] dagcached = False: enable this feature
+        [core] dagcached_min_update_interval = 30 (s):
+            serialized DAGs are updated in DB when a file gets processed by scheduler,
+            to reduce DB write rate, there is a minimal interval of updating serialized DAGs.
+        [scheduler] dag_dir_list_interval = 300 (s):
+            interval of deleting serialized DAGs in DB when the files are deleted, suggest
+            to use a smaller interval such as 60
+
+    It is used by webserver to load dagbags when dagcached=True. Because reading from
+    database is lightweight compared to importing from files, it solves the webserver
+    scalability issue.
+    """
+    __tablename__ = 'serialized_dag'
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    fileloc = Column(String(2000))
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer)
+    data = Column(Text)
+    last_updated = Column(UtcDateTime)
+
+    __table_args__ = (
+        Index('idx_fileloc_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, dag):
+        from airflow.dag.serialization import Serialization
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
+        self.data = Serialization.to_json(dag)
+        self.last_updated = timezone.utcnow()
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        return int(0xFFFF & int(
```

Comment:
What is this magic constant and int conversion? It is fairly opaque so needs a comment explaining what and why we do this.

On python3 at least we can do `int.from_bytes(hashlib.sha1("abc".encode('utf-8')).digest()[-2:], byteorder='big', signed=False)` -- though also: we should probably use more than the last two bytes for this.

Reply:
Will let @coufon answer this one.

Reply:
dag_fileloc_hash is an index of fileloc because fileloc has a length of 2000 in Airflow code, it is too long to be indexed.

I agree to use int.from_bytes(hashlib.sha1("abc".encode('utf-8')).digest()[-2:], byteorder='big', signed=False). It is more clear!

dag_fileloc_hash is not required to be unique. To avoid hash collision, we need to verify the full fileloc anyway, after querying by fileloc. So there is no concern of using 2 bytes in regard of correctness. Do we have other concerns?

Reply:
Are we happy to use the previous version for Py2 ? @ashb @coufon ?

Reply:
Sounds good, yes.